src/scripts/scenes/home.js

```diff
@@ -1,6 +1,7 @@
 import * as THREE from 'three';
 import wallTexture from '../../assets/wall.png';
 import XrScene from './xr-scene';
+import { navigate } from '../router';
```

Comment:
Instead of importing the navigate function, we can set up an event listener in the router to respond to an event we create called 'changeRoom'. The event object that we pass into the navigate function will contain a reference to the scene's room constructor. For example, the home room would add the Home() constructor to the event object.

Here is an example from my personal website:

### Room parent class
```javascript
changeRoom(construct, name) {
        //Remove all associated event listeners when moving to the next room so they don't carry over
        const event = new CustomEvent('changeRoom', {detail : construct});

        const newPath = "/#" + name;
        window.history.pushState({}, newPath, window.location.origin + newPath);

        window.dispatchEvent(event);
    }
```

### Router
```javascript

window.addEventListener('changeRoom', onChangeRoom, true);

. . . 

/**
     * Removes all event listeners associated with current room
     * Stops the current room's rendering loop
     * Creates a new current room from the constructor passed in the event detail
     * Begins render loop for new room.
     * @param {CustomEvent} event 
     */
    onChangeRoom(event) {
        currentRoom.removeEventListeners();
        currentRoom.isActive = false;
        currentRoom = new event.detail(renderer, camera);
        currentRoom.render();
    }
```

### Firing the event
Then for every room, import the room class that it must travel to and whenever you click on a door you can call this:
```javascript
import {anotherRoom} from 'anotherRoom';

. . . 

//Add a field to door mesh to hold constructor for room it links to
room.linkConstructor = anotherRoom;

. . .

this.changeRoom(door.linkConstructor)
```

If you want to see more, here is a link to my source code. The relevant files are src/scripts/index.js, src/scripts/scenes/room.js and src/scripts/scenes/home.js

https://github.com/jonesjonathan/jonesjonathan.github.io/tree/master/src/scripts